Samples/ios/SampleApp/SampleAppiOS/SamplePhotoUtil.m

```diff
@@ -0,0 +1,293 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+#import <MSAL/MSAL.h>
+
+#import "SampleAppErrors.h"
+#import "SamplePhotoUtil.h"
+#import "SampleMSALUtil.h"
+
+@interface SamplePhotoRequest : NSObject
+{
+    NSString *_token;
+}
+
++ (instancetype)requestWithToken:(NSString *)token;
+
+- (void)getPhotoData:(void (^)(NSData *data, NSError *error))photoBlock;
+
+@end
+
+
+@implementation SamplePhotoUtil
+
+
++ (void)getPhoto:(NSString *)token
+           block:(PhotoBlock)photoBlock
+{
+    
+    SamplePhotoRequest *request = [SamplePhotoRequest requestWithToken:token];
+    [request getPhotoData:^(NSData *data, NSError *error) {
+        if (error)
+        {
+            photoBlock(nil, error);
+            return;
+        }
+        
+        if (!data)
+        {
+            [self setLastChecked];
+            NSLog(@"No data returned from graph for photo");
+            photoBlock([UIImage imageNamed:@"no_photo"], nil);
+            return;
+        }
+        
+        UIImage *image = [[UIImage alloc] initWithData:data];
+        if (!image)
+        {
+            photoBlock(nil, [NSError errorWithDomain:SampleAppErrorDomain
+                                                code:SampleAppFailedToMakeUIImageError
+                                            userInfo:nil]);
+            return;
+        }
+        
+        [self cachePhoto:data];
+        
+        
+            photoBlock(image, nil);
+    }];
+}
+
++ (void)getUserPhoto:(PhotoBlock)photoBlock
+{
+    // Start by checking if we have a cached image available for this user
+    UIImage *cached = [self checkPhotoCache];
+    if (cached)
+    {
+        photoBlock(cached, nil);
+        return;
+    }
+    
+    [self getUserPhotoImpl:^(UIImage *photo, NSError *error)
+     {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            photoBlock(photo, error);
+        });
+     }];
+}
+
++ (void)getUserPhotoImpl:(PhotoBlock)photoBlock
+{
+    [SampleMSALUtil acquireTokenSilentForCurrentUser:^(NSString *token, NSError *error)
+     {
+         if (error)
+         {
+             // What an app does on an InteractionRequired error will vary from app to app. More complex apps
+             // will usually want to present a notification to the user in an unobtrusive way (such as on a
+             // status bar in the application UI). Simpler apps, like this one, can jump straight into the
+             // acquireTokenInteractive flow.
+             if ([error.domain isEqualToString:MSALErrorDomain] && error.code == MSALErrorInteractionRequired)
+             {
+                 [SampleMSALUtil acquireTokenInteractiveForCurrentUser:^(NSString *token, NSError *error)
```

Comment:
the nested block here is very small and bottoms out into `-getPhoto:block:` for that reason. All that we have in here is error checking.